src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "InjectLibraryInTracee.h"
+
+#include <absl/base/casts.h>
+#include <absl/strings/match.h>
+#include <absl/strings/str_format.h>
+#include <dlfcn.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+#include <csignal>
+#include <string>
+#include <vector>
+
+#include "AccessTraceesMemory.h"
+#include "AllocateInTracee.h"
+#include "ElfUtils/ElfFile.h"
+#include "ElfUtils/LinuxMap.h"
+#include "MachineCode.h"
+#include "OrbitBase/Logging.h"
+#include "UserSpaceInstrumentation/RegisterState.h"
+
+namespace orbit_user_space_instrumentation {
+
+namespace {
+
+using orbit_elf_utils::ElfFile;
+using orbit_elf_utils::ReadModules;
+
+// Size of the small amount of memory we need in the tracee to write machine code into.
+constexpr uint64_t kScratchPadSize = 1024;
+
+// In certain error conditions the tracee is damaged and we don't try to recover from that. We just
+// abort with a fatal log message. None of these errors are expected to occur in operation
+// obvioulsy. That's what the *OrDie methods below are for.
+void FreeMemoryOrDie(pid_t pid, uint64_t address_code, uint64_t size) {
+  auto result = FreeInTracee(pid, address_code, size);
+  FAIL_IF(result.has_error(), "Unable to free previously allocated memory in tracee: \"%s\"",
+          result.error().message());
+}
+
+void RestoreRegistersOrDie(RegisterState& register_state) {
+  auto result = register_state.RestoreRegisters();
+  FAIL_IF(result.has_error(), "Unable to restore register state in tracee: \"%s\"",
+          result.error().message());
+}
+
+uint64_t GetReturnValueOrDie(pid_t pid) {
+  RegisterState return_value_registers;
+  auto result_return_value = return_value_registers.BackupRegisters(pid);
+  FAIL_IF(result_return_value.has_error(), "Unable to read registers after function called :\"%s\"",
+          result_return_value.error().message());
+  return return_value_registers.GetGeneralPurposeRegisters()->x86_64.rax;
+}
+
+// Execute the code at `address_code`. The code segment has to end with an `int3`.
+void ExecuteOrDie(pid_t pid, RegisterState& original_registers, uint64_t address_code) {
```

Comment:
Shouldn't it be `code_address`?

Reply:
It's pretty much the same I think. The code has an address which is the address of the code.

Reply:
That's the point, in English it's either "address of the code" or "code address", "address code" means something else and it's weird to read.

Reply:
Sorry, I really don't get it. "address of the code" and "address code" are identical to me. 

What happens in my brain when I read "address_code": It's an adress! What might it be the address of? Ah, the code!
A bit like a subscript in a math symbol: F_gravity.

Reply:
Ah, I get it now. For me this is weird and something that I haven't really seen done purposefully, especially not in Orbit.
The name of the variable (excluding special names like `i`) should read like English, where the attributive noun (the noun that acts like an adjective) precedes the noun, hence "code address" = "address of (the) code". "address code" does have a meaning and it's the opposite, that is, "code of the address". The underscore here is not indicating a subscript, but a space.
The same applies to all the `result_` variables in the end, but with "result" it's naturally clearer (and also they're "just" local variables). But yes, personally I would also write those differently.
I can see that I'm not changing your mind but again, I think it's not good practice to sometimes not follow English rules in parameter names (and I don't really get the criterion).

Reply:
To clarify further, here's a question: Why is the function below called `FindFunctionAddress`? Based on your reasoning it should be called `FindAddressFunction`, but of course that would make no sense. So I don't understand why there you almost always apply English rules but then here you don't.

Reply:
Hmm, it really appears totally naturally to me to use it this way around. I would probably not use FindAddressFunction because there is a verb and I would not omit the "Of" then. My code is probably full of these (as you mentioned: result_*). I would honestly never have come up with the idea that anyone would do it differently then adding suffixes to disambiguate different entities. No clue what to make out of this.